File: termux-shared/src/main/java/com/termux/shared/shell/command/result/ResultData.java
package com.termux.shared.shell.command.result;

import androidx.annotation.NonNull;

import com.termux.shared.errors.Error;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class ResultData implements Serializable {

    /**
     * The stdout of command.
     */
    public final StringBuilder stdout = new StringBuilder();

    /**
     * The internal errors list of command.
     */
    public List<Error> errorsList = new ArrayList<>();

    public ResultData() {
    }

    public boolean isStateFailed() {
        if (errorsList != null) {
            for (Error error : errorsList) if (error.isStateFailed())
                return true;
        }
        return false;
    }

    @NonNull
    @Override
    public String toString() {
        return getResultDataLogString(this);
    }

    /**
     * Get a log friendly {@link String} for {@link ResultData} parameters.
     *
     * @param resultData The {@link ResultData} to convert.
     * @return Returns the log friendly {@link String}.
     */
    public static String getResultDataLogString(final ResultData resultData) {
        if (resultData == null)
            return "null";

        return "\n\n" + getErrorsListLogString(resultData);
    }



    public static String getErrorsListLogString(final ResultData resultData) {
        if (resultData == null)
            return "null";
        StringBuilder logString = new StringBuilder();
        if (resultData.errorsList != null) {
            for (Error error : resultData.errorsList) {
                if (error.isStateFailed()) {
                    if (!logString.toString().isEmpty())
                        logString.append("\n");

                }
            }
        }
        return logString.toString();
    }

}

File: termux-shared/src/main/java/com/termux/shared/shell/command/result/ResultData.kt
package com.termux.shared.shell.command.result

import com.termux.shared.errors.Error


class ResultData {
    /**
     * The stdout of command.
     */
    @JvmField
    val stdout: StringBuilder = StringBuilder()

    /**
     * The internal errors list of command.
     */
    private val errorsList: Iterable<Error> = ArrayList()

    val isStateFailed: Boolean
        get() {
            for (error in errorsList) {
                if (error.isStateFailed) return true
            }
            return false
        }

    override fun toString(): String {
        return getResultDataLogString(
            this
        )
    }

    companion object {
        /**
         * Get a log friendly [String] for [ResultData] parameters.
         *
         * @param resultData The [ResultData] to convert.
         * @return Returns the log friendly [String].
         */
        fun getResultDataLogString(resultData: ResultData?): String {
            if (resultData == null) return "null"

            return """
                 
                 
                 ${getErrorsListLogString(resultData)}
                 """.trimIndent()
        }

        private fun getErrorsListLogString(resultData: ResultData?): String {
            if (resultData == null) return "null"
            val logString = StringBuilder()
            for (error in resultData.errorsList) {
                if (error.isStateFailed) {
                    if (!logString.toString().isEmpty()) logString.append("\n")
                }
            }
            return logString.toString()
        }
    }
}
